server/shared/src/main/scala/org/http4s/server/middleware/DefaultHead.scala

```diff
@@ -35,17 +36,38 @@ import org.http4s.Method.HEAD
   * requiring more optimization should implement their own HEAD handler.
   */
 object DefaultHead {
-  def apply[F[_]: Functor, G[_]: Concurrent](http: Http[F, G])(implicit F: MonoidK[F]): Http[F, G] =
+  def apply[F[_]: Functor, G[_]: Applicative](
+      http: Http[F, G]
+  )(implicit F: MonoidK[F]): Http[F, G] =
     Kleisli { req =>
       req.method match {
         case HEAD => http(req) <+> http(req.withMethod(GET)).map(drainBody[G])
         case _ => http(req)
       }
     }
 
-  def httpRoutes[F[_]: Concurrent](httpRoutes: HttpRoutes[F]): HttpRoutes[F] =
+  @deprecated("Use overload with Applicative constraint", "0.23.17")
+  def apply[F[_], G[_]](
+      http: Http[F, G],
+      F: Functor[F],
+      G: Concurrent[G],
+      M: MonoidK[F],
+  ): Http[F, G] =
+    apply(http)(F, G, M)
+
+  def httpRoutes[F[_]: Monad](httpRoutes: HttpRoutes[F]): HttpRoutes[F] =
+    apply(httpRoutes)
+
+  @deprecated("Use overload with Monad constraint", "0.23.17")
+  def httpRoutes[F[_]](httpRoutes: HttpRoutes[F], F: Concurrent[F]): HttpRoutes[F] = {
+    implicit val concurrent = F
     apply(httpRoutes)
+  }
 
-  private[this] def drainBody[G[_]: Concurrent](response: Response[G]): Response[G] =
-    response.pipeBodyThrough(_.interruptWhen[G](Stream(true)).drain)
+  private[this] def drainBody[G[_]](
+      response: Response[G]
+  )(implicit G: Applicative[G]): Response[G] =
+    response.pipeBodyThrough(
+      _.interruptWhen[G](G.pure[Either[Throwable, Unit]](Right(()))).drain
```

Comment:
```suggestion
      _.interruptWhen[G](Either.unit[Throwable].pure[G]).drain
```

Reply:
Does Scala 2.12 contain `Either.unit`? And I adore syntaxes, but I'm in favor of reducing allocations whenever possible within the library.

Reply:
Ah, `Either.unit` is from `cats`, cool. In fence of reducing allocations, it'd be nice to use it.

Reply:
`Either.unit` is provided by Cats via `import cats.syntax.either._`
It expands to `new EitherObjectOps(Either)` but since it is a value class, there should be no allocation involved:
https://github.com/typelevel/cats/blob/b036fb47196d1a788db09eb2bc262071fa3cb928/core/src/main/scala/cats/syntax/either.scala#L358

Moreover, `Either.unit` uses a cached value of `Right(())` internally, hence there should be even less allocations overall:
https://github.com/typelevel/cats/blob/b036fb47196d1a788db09eb2bc262071fa3cb928/core/src/main/scala/cats/syntax/either.scala#L418

Reply:
..in theory, of course :)